src/components/ThemeEngine/StyleEngine.tsx

```diff
@@ -0,0 +1,76 @@
+import { CSS_PREFIX } from '../constants';
+import { kebabCase } from 'lodash';
+import React, { ReactNode, useMemo } from 'react';
+
+import { Root as PortalRoot } from '@radix-ui/react-portal';
+
+/**
+ * Validates a CSS color variable.
+ * @param value The value to validate
+ */
+const isColor = (value: string) => {
+  const s = new Option().style;
+  s.background = value;
+  return s.background !== '';
+};
+
+/**
+ * Adds children to document.head via Radix Portal.
+ * @param children The children to add to document.head
+ */
+const HeadPortal = ({ children }: { children: ReactNode }) => {
+  return (
+    <PortalRoot asChild={true} container={document.head}>
+      {children}
+    </PortalRoot>
+  );
+};
+
+type Styles = {
+  [key: string]: string;
+};
+
+/**
+ * Converts an object of key/value pairs into a CSS string.
+ * Values are validated to ensure they are valid CSS color variables.
+ * Variable names are formated to kebab-case and have --zui appended.
+ * @param styles
+ */
+const mapStylesToCSS = (styles: Styles): string => {
+  return Object.entries(styles)
+    .map(([style, value]) => {
+      // Theme values are currently all colors - check they are valid
+      // colors before adding them to the output styles
+      if (isColor(value)) {
+        return `--${CSS_PREFIX}-${kebabCase(style)}: ${value};`;
+      } else {
+        console.warn('CSS variable not supported', style, value);
+        return;
+      }
+    })
+    .join(' ')
```

Comment:
I think we need to add filter here to remove `undefined` (invalid color) .

```
...
.map(....)
.filter(Boolean)
...
```

Reply:
Hmm yeah probably right, good spot.